sdks/python/apache_beam/io/gcp/experimental/spannerio.py

```diff
@@ -581,3 +644,369 @@ def display_data(self):
+  Batches mutations together.
+  """
+
+  def __init__(self, max_batch_size_bytes):
+    self._max_batch_size_bytes = max_batch_size_bytes
+
+  def start_bundle(self):
+    self._batch = MutationGroup()
+    self._size_in_bytes = 0
+
+  def process(self, element):
+    _max_bytes = self._max_batch_size_bytes
+    mg_size = element.byte_size  # total size of the mutation group.
+
+    if mg_size + self._size_in_bytes > _max_bytes:
+      # Batch is full, output the batch and resetting the count.
+      yield self._batch
+      self._size_in_bytes = 0
+      self._batch = MutationGroup()
+
+    self._batch.extend(element)
+    self._size_in_bytes += mg_size
+
+  def finish_bundle(self):
+    if self._batch is not None:
+      yield window.GlobalWindows.windowed_value(self._batch)
+      self._batch = None
+
+
+@with_input_types(MutationGroup)
+@with_output_types(MutationGroup)
+class _BatchableFilterFn(DoFn):
+  """
+  Filters MutationGroups larger than the batch size to the output tagged with
+  OUTPUT_TAG_UNBATCHABLE.
+  """
+  OUTPUT_TAG_UNBATCHABLE = 'unbatchable'
+
+  def __init__(self, max_batch_size_bytes):
+    self._max_batch_size_bytes = max_batch_size_bytes
+    self._batchable = None
+    self._unbatchable = None
+
+  def process(self, element):
+    if element.primary().operation == 'delete':
+      # As delete mutations are not batchable.
+      yield TaggedOutput(_BatchableFilterFn.OUTPUT_TAG_UNBATCHABLE, element)
+    else:
+      _max_bytes = self._max_batch_size_bytes
+      mg = element
+      mg_size = mg.byte_size
+      if mg_size > _max_bytes:
+        yield TaggedOutput(_BatchableFilterFn.OUTPUT_TAG_UNBATCHABLE, element)
+      else:
+        yield element
+
+
+class _WriteToSpannerDoFn(DoFn):
+
+  def __init__(self, spanner_configuration):
+    self._spanner_configuration = spanner_configuration
+    self._db_instance = None
+    self.batches = Metrics.counter(self.__class__, 'SpannerBatches')
+
+  def setup(self):
+    spanner_client = Client(self._spanner_configuration.project)
+    instance = spanner_client.instance(self._spanner_configuration.instance)
+    self._db_instance = instance.database(
+        self._spanner_configuration.database,
+        pool=self._spanner_configuration.pool)
+
+  def process(self, element):
+    self.batches.inc()
+    with self._db_instance.batch() as b:
+      for m in element:
+        if m.operation == WriteMutation._OPERATION_DELETE:
+          batch_func = b.delete
+        elif m.operation == WriteMutation._OPERATION_REPLACE:
+          batch_func = b.replace
+        elif m.operation == WriteMutation._OPERATION_INSERT_OR_UPDATE:
+          batch_func = b.insert_or_update
+        elif m.operation == WriteMutation._OPERATION_INSERT:
+          batch_func = b.insert
+        elif m.operation == WriteMutation._OPERATION_UPDATE:
+          batch_func = b.update
+        else:
+          raise ValueError("Unknown operation action: %s" % m.operation)
+
+        batch_func(**m.kwargs)
```

Comment:
What is the error handling here?

Say a mutation fails to be written, what will happen to the pipeline?

(The Java code has a option choosing between failing the entire pipeline, or outputting failed mutations to a separate output for later analysis.)

Reply:
Since this write Write mutation code is tightly structured, that would be a very rare case (almost none) that the user will hit that exception. Here just making sure that the operation type should be one of the constant set in the white mutation class. And for now, I haven't applied functionality for the failed mutations to the separate output. Do we need to add that functionality to this version?

Reply:
Sorry, by _"What is the error handling here"_ I am asking what will happen if the Mutation fails to be written to Spanner 
- eg if the Mutation references a table or column that does not exist, or a value is not correct for the column type.

If a failure to write a single mutation would cause the entire pipeline to fail (which I think is the behavior now), then that should at least be documented, and perhaps something that should be introduced in a future version.

Reply:
Sure, i'll update the code!